# Interal Optimisation Tutorial

## Setup

The `IntervalOptimisation.jl` package can be installed with

In [1]:
using Pkg; Pkg.add("IntervalOptimisation")

   Updating registry at `C:\Users\lucaa\.julia\registries\General`
   Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Resolving package versions...
   Updating `C:\Users\lucaa\.julia\environments\v1.4\Project.toml`
 [no changes]
   Updating `C:\Users\lucaa\.julia\environments\v1.4\Manifest.toml`
 [no changes]


Once the package is installed, it can be imported. Note that you will need also the `IntervalArithmetic.jl` package.

In [2]:
using IntervalArithmetic, IntervalOptimisation

## Function Minimisation

The package two main functions are `minimise` and `maximise`. Here we will use `minimise` as example, however `maximise` behaves in an analog way.
The main syntax is
```
minimise(f, X, tol=1e-3),
```

where $f:\R^n↦\R$ is the function to minimize, $X$ is the interval, or interval box, over which we minimize the function.
For example, let's consider the function $f(x)=(x-3)^2$ and let us find its global minimum over its whole domain

In [3]:
minimise(x->(x-3)^2, -∞..∞)

([0, 7.39292e-09], IntervalArithmetic.Interval{Float64}[[2.99964, 3.00019]])

The first value of the results tells us that the minimum value of the function is in the interval $[0, 7.39292e-09]$. The second value tells us that
this minimum is achieved in the interval $[2.99964, 3.00019]$. If we want to narrow down this interval, we can use a smaller tollerance

In [4]:
minVal, xmin = minimise(x->(x-3)^2, -∞..∞, tol=1e-9)
xmin, [diam(x) for x in xmin]

(IntervalArithmetic.Interval{Float64}[[2.99999, 3.00001]], [5.109304090922251e-10])

Now the diameter of the minimiser is only $10^{-10}$ and the minimum is guaranteed to be in that interval.

## Multivariate Function Minimisation

The package can also be used to minimise multivariate functions. Now, the function $f$ will take an array, and $X$ will be an interval box of dimension $n$.
As an example, let us find the minimum of the paraboloid $z=(x-3)^2+(y-5)^22$ over the box $[-100, 100]×[-100, 100]$. First let's define the function

In [5]:
paraboloid(x) = (x[1]-3)^2 + (x[2]-3)^2

paraboloid (generic function with 1 method)

and our interval box

In [6]:
X = IntervalBox(-100..100, 2)

[-100, 100] × [-100, 100]

now we can obtain the minimum

In [7]:
zmin, xmin = minimise(paraboloid, X, tol=1e-10)

([0, 4.24526e-22], IntervalArithmetic.IntervalBox{2,Float64}[[2.99999, 3.00001] × [2.99999, 3.00001]])

the position of the minimum has been found with an accuracy of

In [8]:
[diam(x) for x in xmin]

1-element Array{Float64,1}:
 9.154810243217071e-11

### Griewank Function Minimisation

The $n$-dimensional Griewank function is defined as
$$ G_n(\mathbf{x})=1+\frac{1}{4000}\sum_{i=1}^n x_i^2 -∏_{i=1}^n \cos\left(\frac{x_i}{\sqrt{i}}\right),$$

for example the 1-dimensional Griewank function is

$$ G_1(x) = 1+\frac{x^2}{4000}-\cos(x) $$.

and it is commonly used to test optimisation algorithms. This function has the property to have several regularly distributed local minima, but only one global minimum at the origin. Let's define our function

In [9]:
G(X) = 1 + sum(abs2, X) / 4000 - prod( cos(X[i] / √i) for i in 1:length(X) )

G (generic function with 1 method)

Now let's verify our package finds the minima in several dimensions

In [10]:
for N in (1,2, 10, 20, 50)
    res, xmin = minimise(G, IntervalBox(-600..600, N))
    @show N, res, diam(xmin[1])
end

(N, res, diam(xmin[1])) = (1, [-0, 0], 0.000541404722891739)
(N, res, diam(xmin[1])) = (2, [-0, 0], 0.000541404722891739)
(N, res, diam(xmin[1])) = (10, [-0, 0], 0.000541404722891739)
(N, res, diam(xmin[1])) = (20, [-0, 0], 0.000541404722891739)
(N, res, diam(xmin[1])) = (50, [-0, 0], 0.000541404722891739)


---

*This notebook was generated using [Literate.jl](https://github.com/fredrikekre/Literate.jl).*